diff -ur openfst-1.8.3-orig/src/include/fst/impl-to-fst.h openfst-1.8.3/src/include/fst/impl-to-fst.h
--- openfst-1.8.3-orig/src//include/fst/impl-to-fst.h	2024-01-11 22:32:17
+++ openfst-1.8.3/src/include/fst/impl-to-fst.h	2024-07-28 12:20:05
@@ -127,7 +127,7 @@
   // Returns a ref-counted smart poiner to the implementation.
   std::shared_ptr<Impl> GetSharedImpl() const { return impl_; }

-  bool Unique() const { return impl_.unique(); }
+  bool Unique() const { return impl_.use_count() == 1; }

   void SetImpl(std::shared_ptr<Impl> impl) { impl_ = std::move(impl); }

diff -ur openfst-1.8.3-orig/src/include/fst/symbol-table.h openfst-1.8.3/src/include/fst/symbol-table.h
--- openfst-1.8.3-orig/src/include/fst/symbol-table.h	2024-07-28 12:13:29
+++ openfst-1.8.3/src/include/fst/symbol-table.h	2024-07-28 12:11:37
@@ -522,7 +522,7 @@

  private:
   void MutateCheck() {
-    if (impl_.unique() || !impl_->IsMutable()) return;
+    if (impl_.use_count() == 1 || !impl_->IsMutable()) return;
     std::unique_ptr<internal::SymbolTableImplBase> copy = impl_->Copy();
     CHECK(copy != nullptr);
     impl_ = std::move(copy);

diff -ur openfst-1.8.3-orig/src/include/fst/log.h openfst-1.8.3/src/include/fst/log.h
--- openfst-1.8.3-orig/src/include/fst/log.h	2022-02-03 17:48:58.000000000 +0100
+++ openfst-1.8.3/src/include/fst/log.h	2022-09-08 10:19:17.000000000 +0200
@@ -29,18 +29,20 @@
 
 DECLARE_int32(v);
 
+extern std::ostream *FST_FLAGS_outputstream;
+
 class LogMessage {
  public:
   explicit LogMessage(std::string_view type) : fatal_(type == "FATAL") {
-    std::cerr << type << ": ";
+    (FST_FLAGS_outputstream ? *FST_FLAGS_outputstream : std::cerr) << type << ": ";
   }
 
   ~LogMessage() {
-    std::cerr << std::endl;
+    (FST_FLAGS_outputstream ? *FST_FLAGS_outputstream : std::cerr) << std::endl;
     if (fatal_) exit(1);
   }
 
-  std::ostream &stream() { return std::cerr; }
+  std::ostream &stream() { return (FST_FLAGS_outputstream ? *FST_FLAGS_outputstream : std::cerr); }
 
  private:
   bool fatal_;
diff -ur openfst-1.8.3-orig/src/lib/util.cc openfst-1.8.3/src/lib/util.cc
--- openfst-1.8.3-orig/src/lib/util.cc	2022-02-03 17:48:58.000000000 +0100
+++ openfst-1.8.3/src/lib/util.cc	2022-09-08 09:18:28.000000000 +0200
@@ -37,6 +37,8 @@
             "FST errors are fatal; o.w. return objects flagged as bad: "
             "e.g., FSTs: kError property set, FST weights: not a Member()");
 
+std::ostream *FST_FLAGS_outputstream = nullptr;
+
 namespace fst {
 
 std::optional<int64_t> ParseInt64(std::string_view s, int base) {
